packages/ui-a11y-utils/src/KeyboardFocusRegion.ts

```diff
@@ -160,6 +167,45 @@ class KeyboardFocusRegion {
     }
   }
 
+  tryFocusOnPositionedElement(
+    positionContainer: Element,
+    focusElement: HTMLElement,
+    tries = 3
+  ) {
+    // If element is positioned by `ui-position`
+    // we may need to wait for it to appear in
+    // correct place before triggering focus.
+    // Otherwise it will cause a scroll to the top of the page.
+
+    if (tries <= 0) {
+      return
+    }
+
+    const positionRect = positionContainer.getBoundingClientRect()
+    const docRect = document.documentElement.getBoundingClientRect()
+
+    // Check if element is outside of visible part of page (including parts hidden by scroll).
+    const isHiddenOutside =
+      positionRect.top > docRect.bottom ||
+      positionRect.right < docRect.left ||
+      positionRect.bottom < docRect.top ||
+      positionRect.left > docRect.right
+
+    if (!isHiddenOutside) {
+      focusElement.focus()
+    } else {
+      this._raf.push(
+        requestAnimationFrame(() => {
+          this.tryFocusOnPositionedElement(
+            positionContainer,
+            focusElement,
+            tries - 1
+          )
+        })
```

Comment:
I tried `MutationObserver` and `IntersectionObserver` but this simple solution works the best